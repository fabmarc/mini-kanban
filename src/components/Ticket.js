import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import { useDrag } from 'react-dnd';
import PropTypes from 'prop-types';

import TextBox from './TextBox';
import Button from './Button';
import Title from './Title';

const Wrapper = styled.div`
  width: 100%;
  height: 58px;
  display: flex;
  align-items: center;
  border-radius: 5px;
  border: 1px solid #c0c0c0;
  margin-bottom: 8px;
  opacity: ${props => props.opacity || 1};
  animation-name: ticket;
  animation-duration: 2s;
  background-color: #fffef0;
`;

const inputRef = React.createRef();

function Ticket({ value = {}, onChange, onDelete }) {
  const origDescription = value.description || '';
  const [editTicket, setEditTicket] = useState(false);
  const [description, setDescription] = useState(origDescription);

  useEffect(() => {
    if (editTicket && inputRef.current) inputRef.current.focus();
  }, [editTicket]);

  const [{ isDragging }, drag] = useDrag({
    item: { type: 'ticket', ticket: value },
    collect: monitor => ({
      isDragging: !!monitor.isDragging(),
    }),
  });

  const saveTicket = () => {
    if (!description) setDescription(origDescription);
    else if (description !== origDescription) {
      if (onChange) onChange(value, description);
      setDescription(description);
    }
    setEditTicket(false);
  }

  const handleValueChange = (event) => {
    setDescription(event.target.value);
  }

  const handleDeleteClick = () => {
    if (onDelete) onDelete(value);
  }

  const handleDoubleClick = () => {
    setEditTicket(true);
  }

  if (editTicket) {
    return (
      <TextBox
        ref={inputRef}
        onBlur={saveTicket}
        value={description}
        onChange={handleValueChange}
      />
    );
  }
  return (
    <Wrapper
      ref={drag}
      opacity={isDragging ? 0.2 : 1}
      onDoubleClick={handleDoubleClick}
    >
      <Title>{origDescription || '(no description)'}</Title>
      <Button onClick={handleDeleteClick}>-</Button>
    </Wrapper>
  );
}

Ticket.propTypes = {
  value: PropTypes.shape({
    id: PropTypes.number.isRequired,
    description: PropTypes.string.isRequired,
    status: PropTypes.string.isRequired,
  }),
}

export default Ticket;
