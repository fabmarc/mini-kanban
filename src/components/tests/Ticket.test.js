import React from 'react';
import { shallow, mount } from 'enzyme';
import HTML5Backend from 'react-dnd-html5-backend';
import renderer from 'react-test-renderer';
import { DndProvider } from 'react-dnd';
import Ticket from '../Ticket';

describe('<Ticket>', () => {
  it('renders without crashing', () => {
    shallow(
      <DndProvider backend={HTML5Backend}>
        <Ticket />
      </DndProvider>
    );
  });

  it('renders default description', () => {
    const wrapper = mount(
      <DndProvider backend={HTML5Backend}>
        <Ticket />
      </DndProvider>
    );
    expect(wrapper.find('div').at(1).contains('(no description)')).toBe(true);
  });

  it('renders in edit mode', () => {
    const wrapper = mount(
      <DndProvider backend={HTML5Backend}>
        <Ticket />
      </DndProvider>
    );
    wrapper.find('div').first().simulate('dblclick');
    expect(wrapper.find('input').length).toEqual(1);
  });

  it('triggers onChange event', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <DndProvider backend={HTML5Backend}>
        <Ticket onChange={onChange} />
      </DndProvider>
    );
    wrapper.find('div').first().simulate('dblclick');
    wrapper.find('input').simulate('change', { target: { value: 'Test' } });
    wrapper.find('input').simulate('blur');
    expect(onChange.mock.calls.length).toEqual(1);
  });

  it('triggers onDelete event', () => {
    const onDelete = jest.fn();
    const wrapper = mount(
      <DndProvider backend={HTML5Backend}>
        <Ticket onDelete={onDelete} />
      </DndProvider>
    );
    wrapper.find('button').simulate('click');
    expect(onDelete.mock.calls.length).toEqual(1);
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <DndProvider backend={HTML5Backend}>
          <Ticket />
        </DndProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
