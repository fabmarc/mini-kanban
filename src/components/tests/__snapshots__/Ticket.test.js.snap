// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Ticket> renders correctly 1`] = `
<div
  className="sc-bxivhb gGKUWd"
  onDoubleClick={[Function]}
  opacity={1}
>
  <div
    className="sc-htpNat cmbrzm"
  >
    (no description)
  </div>
  <button
    className="sc-bwzfXH eUNjes"
    onClick={[Function]}
  >
    -
  </button>
</div>
`;

exports[`renders correctly 1`] = `
<div
  className="sc-bxivhb gGKUWd"
  onDoubleClick={[Function]}
  opacity={1}
>
  <div
    className="sc-htpNat cmbrzm"
  >
    (no description)
  </div>
  <button
    className="sc-bwzfXH eUNjes"
    onClick={[Function]}
  >
    -
  </button>
</div>
`;
